File: extension.ts

/**
 * Communicate with nRF24 devices
 */
namespace radio24 {
    let onReceivedBufferHandler: (receivedBuffer: Buffer) => void;

    let initialized = false;
    function init() {
        if (! initialized) {
            initialized = true;
            onDataReceived(handleDataReceived);
        }
    }

    function handleDataReceived() {
        let buffer: Buffer = readRawPacket();

        while (buffer) {
            if (onReceivedBufferHandler) {
                onReceivedBufferHandler(buffer);
            }
            buffer = readRawPacket();
        }
    }

    /**
     * Register receive handler
     */
    //% blockId=radio24_on_receive block="on radio received"
    //% useLog="radio24.onDataBufferReceived" draggableParameters=reporter
    export function onReceivedBuffer(cb: (receivedBuffer: Buffer) => void) {
        init();
        onReceivedBufferHandler = cb;
    }

    /**
     * Send data
     * 
     * @param msg Buffer to send
     */
    //% advanced=true
    export function sendBuffer(msg: Buffer) {
        sendRawPacket(msg);
    }

    /**
     * Send string
     */
    //% blockId=radio24_send_string block="Send $str"
    export function sendString(str: string) {
        let buf = control.createBufferFromUTF8(str);
        sendRawPacket(buf);
    }

    /**
     * Send raw packet
     * 
     * @param buf Buffer to send
     */
    //% shim=radio24::sendRawPacket
    export function sendRawPacket(buf: Buffer) : void {
        console.log("Sending buffer");
        return;
    }

    /**
     *  Enable radio
     */
    //% blockId=radio24_enable block="Activate radio with band=$band, power=$power"
    //% weight=10
    //% band.min=0 band.max=100 power.min=0 power.max=7
    //% shim=radio24::enable
    export function enable(band?: uint8, power?: uint8): void {
        console.log("Radio enabled");
        return;
    }

    /**
     * Set radio group
     */
    //% blockId=radio24_set_group block="Set group to $group"
    //% group.min=0 group.max=255
    //% shim=radio24::setGroup
    export function setGroup(group: uint8): void {
        console.log("Setting radio group");
        return;
    }

    /**
     * Ping test
     */
    //% blockId=radio24_ping block="Run extension test"
    //% shim=radio24::ping
    export function ping(): void {
        basic.showString("SIM");
        return;
    }
}

File: extension.cpp

#include "pxt.h"
#include <cstdint>
#include "mbed.h"

using namespace pxt;

enum Mode {
    //% block="1Mbps"
    Mode_1Mbps = 0,
    //% block="2Mbps"
    Mode_2Mbps,
    //% block="250Kbps"
    Mode_250Kbps
};

//% color="#AA278D" icon="\uf0a4"
namespace radio24 {
    /**
     * Internal use only. Receive handler.
     */
    //% blockId=radio24_data_received_event block="radio on data received"
    //% deprecated=true blockHidden=true
    void onDataReceived(Action body) {
        registerWithDal(MICROBIT_ID_RADIO, MICROBIT_RADIO_EVT_DATAGRAM, body);
        uBit.radio.datagram.recv();
    }

    //%
    void enable(uint8_t band, uint8_t power) {
        uBit.radio.enable();
        uBit.radio.setTransmitPower(power);
        uBit.radio.setFrequencyBand(band);

        // Make it compatible with nRF24
        NRF_RADIO->PCNF0 = 0x00000006; // on-air LENGTH field length of 6bit
        NRF_RADIO->PCNF1 = 0x01040000 | 32; // no whitening, big endian, 32B payload
    	NRF_RADIO->DATAWHITEIV = 0x00;
        NRF_RADIO->MODE = Mode_1Mbps;
    }

    //%
    void setGroup(uint8_t group) {
        uBit.radio.setGroup(group);
    }

    /**
     * Read buffer data
     * 
     * @param buf Buffer to send
     */
    //%
    Buffer readBuffer() {
        auto p = uBit.radio.datagram.recv();

        if (p == PacketBuffer::EmptyPacket)
            return NULL;
        return mkBuffer(p.getBytes(), p.length());
    }

    //%
    void sendBuffer(Buffer msg) {
        char ch = "0123456789ABCDEF"[msg->length];
        uBit.serial.putc(ch);
        uBit.radio.datagram.send(msg->data, msg->length);
    }

    //%
    void ping() {
        uBit.display.scrollAsync("DEV");
    }
}
